uaclient/cli.py

```diff
@@ -1222,22 +1223,9 @@ def _detach(cfg: config.UAConfig, assume_yes: bool) -> int:
         if ret:
             to_disable.append(ent)
 
-    """
-    We will make sure that services without dependencies are disabled first
-    PS: This will only work because we have only three services with reverse
-    dependencies:
-    * ros: ros-updates
-    * esm-infra: ros, ros-updates
-    * esm-apps: ros, ros-updates
-
-    Therefore, this logic will guarantee that we will always disable ros and
-    ros-updates before diabling the esm services. If that dependency chain
-    change, this logic won't hold anymore and must be properly fixed.
-
-    More details can be seen here:
-    https://github.com/canonical/ubuntu-advantage-client/issues/1831
-    """
-    to_disable.sort(key=lambda ent: len(ent.dependent_services))
+    to_disable.sort(
+        key=lambda ent: entitlements_disable_order(cfg).index(ent.name)
```

Comment:
We should only call `entitlements_disable_order` once by moving it outside of the lambda.

E.g.
```
    disable_order = entitlements_disable_order(cfg)
    to_disable.sort(
        key=lambda ent: disable_order.index(ent.name)
    )
```

Reply:
Alternatively we might be able to replace the above for loop that creates the `to_disable` list to avoid needing a sort later.

Something like:
```diff
     to_disable = []
-    for ent_cls in entitlements.ENTITLEMENT_CLASSES:
+    for name in entitlements_disable_order(cfg):
+        try:
+            ent_cls = entitlement_factory(cfg, name)
+        except EntitlementNotFoundError
+            continue
         ent = ent_cls(cfg=cfg, assume_yes=assume_yes)
         # For detach, we should not consider that a service
         # cannot be disabled because of dependent services,
         # since we are going to disable all of them anyway
         ret, _ = ent.can_disable(ignore_dependent_services=True)
         if ret:
             to_disable.append(ent)

-    to_disable.sort(
-        key=lambda ent: entitlements_disable_order(cfg).index(ent.name)
-    )
```